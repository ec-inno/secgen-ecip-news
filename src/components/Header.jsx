import React from 'react';
import { Link } from 'gatsby';

import logoPaths from '../utils/logoPaths';
import { map as languageMap, defaultLangKey } from '../languages';

import SiteName from './SiteName';
import LanguageListOverlay from './LanguageList/LanguageListOverlayWithContext';
import LanguageSelector from './LanguageSelector';

const Header = ({ languages, location, contentTranslations }) => {
  let items = [];
  let logo = logoPaths[defaultLangKey];
  let opensOverlay = false;

  const { pathname } = location;
  const pathParts = pathname.split('/').filter(p => p);
  const langcodeCurrent = pathParts.shift();
  const urlPath = pathParts.join('/');

  // Change logo to a language-specific one, if applicable.
  if (langcodeCurrent !== defaultLangKey) {
    logo = logoPaths[langcodeCurrent];
  }

  const translationSet = contentTranslations.find(contentTranslation => {
    return (
      contentTranslation.path.alias === `/${urlPath}` &&
      contentTranslation.path.langcode === langcodeCurrent
    );
  });

  // Are there translations for this content page?
  if (
    translationSet &&
    translationSet.translations &&
    translationSet.translations.length
  ) {
    // Add translations.
    items = translationSet.translations.map(translation => {
      // translation is a structure of data coming from jsonapi about a pathauto alias.
      // Because gatsby has already created content matching this path, we only need to reformat the information.
      const { alias, langcode } = translation;

      return {
        href: `/${langcode}${alias}`,
        lang: langcode,
        label: languageMap[langcode],
      };
    });

    // Add currently active language.
    items.push({
      href: pathname,
      lang: langcodeCurrent,
      label: languageMap[langcodeCurrent],
      isActive: true,
    });

    // Enable overlay.
    opensOverlay = true;
  }
  // The page is based on drupal content, not translations.
  else if (!translationSet && pathParts.length > 1) {
    opensOverlay = false;
  }
  // The page is programatic: /news, /about, etc. Translation pattern is predictable.
  else {
    items = languages.map(language => {
      const href = urlPath
        ? `/${language.lang}/${urlPath}`
        : `/${language.lang}`;
      const isActive = href.includes(langcodeCurrent);

      return {
        href,
        isActive,
        ...language,
      };
    });

    // Enable the overlay.
    opensOverlay = true;
  }

  // Correct order.
  items = items.sort((a, b) =>
    a.lang < b.lang ? -1 : a.lang > b.lang ? 1 : 0
  );

  return (
    <>
      <header className="ecl-site-header">
        <div className="ecl-site-header__container ecl-container">
          <div className="ecl-site-header__banner">
            <Link
              className="ecl-link ecl-link--standalone"
              to={`/${langcodeCurrent}`}
              aria-label="European Union"
            >
              <img
                alt="European Union logo"
                title="European Union"
                className="ecl-site-header__logo-image"
                src={logo}
              />
            </Link>
            <LanguageSelector
              code={langcodeCurrent}
              name={languageMap[langcodeCurrent]}
              href="#"
              opensOverlay={opensOverlay}
            />
          </div>
        </div>
        <SiteName currentLanguage={langcodeCurrent} />
      </header>
      <LanguageListOverlay
        closeLabel="Close"
        title="Select your language"
        items={items}
      />
    </>
  );
};

export default Header;
