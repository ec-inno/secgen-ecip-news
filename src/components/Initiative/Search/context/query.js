import React from 'react';

import { defaultLangKey } from '../../../../../languages';

const queryInit = {
  filters: {},
  status: 'ALL',
  language: defaultLangKey,
  pagination: '0/10',
};

const queryReducer = (state, action) => {
  switch (action.type) {
    case 'paginate': {
      return {
        ...state,
        pagination: action.pagination,
      };
    }

    case 'changeStatus': {
      return {
        ...state,
        pagination: queryInit.pagination,
        status: action.status,
      };
    }

    case 'changeLanguage': {
      return {
        ...state,
        pagination: queryInit.pagination,
        language: action.language,
      };
    }

    case 'changeFilter': {
      return {
        ...state,
        pagination: queryInit.pagination,
        filters: {
          ...state.filters,
          [action.filter]: action.filterValue,
        },
      };
    }

    case 'setFilters': {
      return {
        ...state,
        pagination: queryInit.pagination,
        filters: {
          ...state.filters,
          ...action.filters,
        },
      };
    }

    case 'removeFilter': {
      const { filters } = state;
      delete filters[action.filter];

      return {
        ...state,
        pagination: queryInit.pagination,
        filters: {
          ...filters,
        },
      };
    }

    case 'reset':
    default: {
      return queryInit;
    }
  }
};

const Context = React.createContext();

export { queryInit, queryReducer, Context as default };
