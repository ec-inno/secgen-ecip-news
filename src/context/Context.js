import React from 'react';

const initialState = {
  hideOverlay: true,
};

export const SET_LANGUAGE_OVERLAY_VISIBILITY =
  'SET_LANGUAGE_OVERLAY_VISIBILITY';

const reducer = (state, action) => {
  switch (action.type) {
    case SET_LANGUAGE_OVERLAY_VISIBILITY: {
      return Object.assign({}, state, {
        hideOverlay: action.hideOverlay,
      });
    }

    default: {
      return initialState;
    }
  }
};

const Context = React.createContext();

const ContextProviderWithReducer = ({ children }) => {
  const [store, dispatch] = React.useReducer(reducer, initialState);

  return (
    <Context.Provider value={{ store, dispatch }}>{children}</Context.Provider>
  );
};

export { ContextProviderWithReducer, Context as default };
