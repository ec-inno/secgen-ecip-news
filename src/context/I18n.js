import React from 'react';

const I18n = React.createContext();

export default I18n;
