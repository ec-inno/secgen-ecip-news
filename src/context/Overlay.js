import React, { useContext } from 'react';

const Context = React.createContext();

const useOverlayContext = () => useContext(Context);

export { useOverlayContext, Context as default };
