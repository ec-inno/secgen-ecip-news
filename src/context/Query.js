import React, { useContext } from 'react';

import { defaultLangKey } from '../../languages';

const queryInit = {
  filters: {},
  status: 'ALL',
  language: defaultLangKey,
  pagination: '0/10',
};

const queryReducer = (state, action) => {
  switch (action.type) {
    case 'paginate': {
      return {
        ...state,
        pagination: action.pagination,
      };
    }

    case 'changeLanguage': {
      return {
        ...state,
        pagination: queryInit.pagination,
        language: action.language,
      };
    }

    case 'setFilters': {
      return {
        ...state,
        pagination: queryInit.pagination,
        filters: {
          ...state.filters,
          ...action.filters,
        },
      };
    }

    case 'unsetFilter': {
      const { filters } = state;
      delete filters[action.filter];

      return {
        ...state,
        pagination: queryInit.pagination,
        filters: {
          ...filters,
        },
      };
    }

    case 'reset':
    default: {
      return queryInit;
    }
  }
};

const Context = React.createContext();

const useQueryContext = () => useContext(Context);

export { queryInit, queryReducer, Context as default, useQueryContext };
