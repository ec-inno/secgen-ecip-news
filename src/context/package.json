{
  "name": "@eci/context",
  "version": "0.1.0",
  "description": "",
  "private": true,
  "dependencies": {
    "@ecl/eu-preset-website": "2.8.0",
    "react": "16.9.0"
  }
}
