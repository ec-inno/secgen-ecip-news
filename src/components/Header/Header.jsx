import React from 'react';
import { useTranslation } from 'react-i18next';
import { Link } from 'gatsby';

import logoPaths from '@eci/utils/logoPaths';

import { languages } from '../../../languages';
import { useI18nContext } from '../../context/I18n';

import SiteName from '../SiteName';
import LanguageListOverlay from '../LanguageList/LanguageListOverlayWithContext';
import LanguageSelector from '../LanguageSelector';

// Create a map of language code to language label.
const languageMap = languages.reduce((obj, item) => {
  obj[item.lang] = item.label;
  return obj;
}, {});

const Header = () => {
  const { locale, location } = useI18nContext();
  const { t } = useTranslation();
  const logo = logoPaths[locale];

  let urlPath = '';
  let pathParts = [];

  if (location && location.pathname) {
    pathParts = location.pathname.split('/').filter(p => p);
  }

  // Current logic of language switcher: change language prefix.
  // This means that we add current in order to make paths for the translations.
  pathParts.shift();

  urlPath = pathParts.join('/');

  if (location.hash) {
    urlPath += location.hash;
  }

  const items = languages.map(item => {
    const href = urlPath ? `/${item.lang}/${urlPath}` : `/${item.lang}`;
    const isActive = href.startsWith(`/${locale}`);

    return {
      href,
      isActive,
      ...item,
    };
  });

  return (
    <>
      <header className="ecl-site-header">
        <div className="ecl-site-header__container ecl-container">
          <div className="ecl-site-header__banner">
            <Link
              className="ecl-link ecl-link--standalone"
              to={`/${locale}`}
              aria-label={t('European Union')}
            >
              <img
                alt={`${t('European Union')} logo`}
                title={t('European Union')}
                className="ecl-site-header__logo-image"
                src={logo}
              />
            </Link>
            <LanguageSelector
              code={locale}
              name={languageMap[locale]}
              href="#"
            />
          </div>
        </div>
        <SiteName />
      </header>
      <LanguageListOverlay
        closeLabel={t('Close')}
        title={t('Select your language')}
        items={items}
      />
    </>
  );
};

export default Header;
