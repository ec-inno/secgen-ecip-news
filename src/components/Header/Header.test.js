import React from 'react';
import renderer from 'react-test-renderer';

import * as I18nContext from '../../context/I18n';
import * as OverlayContext from '../../context/Overlay';
import Header from './Header';

describe('Header', () => {
  it('renders correctly', () => {
    jest.spyOn(I18nContext, 'useI18nContext').mockImplementation(() => ({
      locale: 'en',
      location: { pathname: '/en/how-it-works' },
    }));

    jest.spyOn(OverlayContext, 'useOverlayContext').mockImplementation(() => ({
      overlayIsHidden: true,
      setOverlayIsHidden: () => true,
    }));

    const tree = renderer.create(<Header />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
