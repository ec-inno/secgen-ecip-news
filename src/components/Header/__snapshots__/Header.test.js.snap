// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
Array [
  <header
    className="ecl-site-header"
  >
    <div
      className="ecl-site-header__container ecl-container"
    >
      <div
        className="ecl-site-header__banner"
      >
        <a
          aria-label="European Union"
          className="ecl-link ecl-link--standalone"
          href="/en"
        >
          <img
            alt="European Union logo"
            className="ecl-site-header__logo-image"
            src="file"
            title="European Union"
          />
        </a>
        <div
          className="ecl-site-header__selector"
        >
          <a
            className="ecl-link ecl-link--standalone"
            data-ecl-language-selector={true}
            href="#"
            onClick={[Function]}
          >
            English
            <span
              className="ecl-site-header__language-icon"
            >
              <svg
                aria-hidden="true"
                className="ecl-icon ecl-icon--m"
                focusable="false"
              >
                
                
                <use
                  xlinkHref="file#general--language"
                />
              </svg>
              <span
                className="ecl-site-header__language-code"
              >
                en
              </span>
            </span>
          </a>
        </div>
      </div>
    </div>
    <div
      className="ecl-container ecl-u-pb-m ecl-u-type-xl ecl-u-type-bold"
    >
      <span>
        European
      </span>
      <span
        className="ecl-u-ml-2xs"
      >
        Citizens'
      </span>
      <span
        className="ecl-u-ml-2xs"
      >
        Initiative
      </span>
    </div>
  </header>,
  <div
    data-focus-guard={true}
    style={
      Object {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={0}
  />,
  <div
    data-focus-guard={true}
    style={
      Object {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={1}
  />,
  <div
    data-focus-lock-disabled={false}
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <div
      className="ecl-language-list ecl-language-list--overlay"
      data-ecl-language-list-overlay={true}
      hidden={true}
      onKeyDown={[Function]}
    >
      <div
        className="ecl-language-list__container ecl-container"
      >
        <div
          className="ecl-row"
        >
          <div
            className="ecl-language-list__close ecl-col-12 ecl-col-lg-8 ecl-offset-lg-2"
          >
            <button
              className="ecl-language-list__close-button ecl-button ecl-button--ghost"
              data-ecl-language-list-close={true}
              disabled={false}
              onClick={[Function]}
              type="submit"
            >
              <span
                className="ecl-button__container"
              >
                <span
                  className="ecl-button__label"
                  data-ecl-label={true}
                >
                  Close
                </span>
                <svg
                  aria-hidden="true"
                  className="ecl-button__icon ecl-button__icon--after ecl-icon ecl-icon--s"
                  data-ecl-icon={true}
                  focusable="false"
                >
                  
                  
                  <use
                    xlinkHref="file#ui--close"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            className="ecl-language-list__title ecl-col-12 ecl-col-lg-8 ecl-offset-lg-2"
          >
            <svg
              aria-hidden="true"
              className="ecl-language-list__title-icon ecl-icon ecl-icon--m"
              focusable="false"
            >
              
              
              <use
                xlinkHref="file#general--generic-lang"
              />
            </svg>
            Select your language
          </div>
        </div>
        <div
          className="ecl-row"
        >
          <div
            className="ecl-language-list__column ecl-col-12 ecl-col-lg-4 ecl-offset-lg-2"
          >
            <ul
              className="ecl-language-list__list"
            >
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/bg/how-it-works"
                  hrefLang="bg"
                  lang="bg"
                  onClick={[Function]}
                  rel="alternate"
                >
                  български
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/es/how-it-works"
                  hrefLang="es"
                  lang="es"
                  onClick={[Function]}
                  rel="alternate"
                >
                  español
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/cs/how-it-works"
                  hrefLang="cs"
                  lang="cs"
                  onClick={[Function]}
                  rel="alternate"
                >
                  čeština
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/da/how-it-works"
                  hrefLang="da"
                  lang="da"
                  onClick={[Function]}
                  rel="alternate"
                >
                  dansk
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/de/how-it-works"
                  hrefLang="de"
                  lang="de"
                  onClick={[Function]}
                  rel="alternate"
                >
                  Deutsch
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/et/how-it-works"
                  hrefLang="et"
                  lang="et"
                  onClick={[Function]}
                  rel="alternate"
                >
                  eesti
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/el/how-it-works"
                  hrefLang="el"
                  lang="el"
                  onClick={[Function]}
                  rel="alternate"
                >
                  ελληνικά
                </a>
              </li>
              <li
                className="ecl-language-list__item ecl-language-list__item--is-active"
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link ecl-link--icon ecl-link--icon-after"
                  href="/en/how-it-works"
                  hrefLang="en"
                  lang="en"
                  onClick={[Function]}
                  rel="alternate"
                >
                  <span
                    className="ecl-link__label"
                  >
                    English
                  </span>
                   
                  <svg
                    aria-hidden="true"
                    className="ecl-link__icon ecl-icon ecl-icon--xs"
                    focusable="false"
                  >
                    
                    
                    <use
                      xlinkHref="file#ui--check"
                    />
                  </svg>
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/fr/how-it-works"
                  hrefLang="fr"
                  lang="fr"
                  onClick={[Function]}
                  rel="alternate"
                >
                  français
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/ga/how-it-works"
                  hrefLang="ga"
                  lang="ga"
                  onClick={[Function]}
                  rel="alternate"
                >
                  Gaeilge
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/hr/how-it-works"
                  hrefLang="hr"
                  lang="hr"
                  onClick={[Function]}
                  rel="alternate"
                >
                  hrvatski
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/it/how-it-works"
                  hrefLang="it"
                  lang="it"
                  onClick={[Function]}
                  rel="alternate"
                >
                  italiano
                </a>
              </li>
            </ul>
          </div>
          <div
            className="ecl-language-list__column ecl-col-12 ecl-col-lg-4"
          >
            <ul
              className="ecl-language-list__list"
            >
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/lv/how-it-works"
                  hrefLang="lv"
                  lang="lv"
                  onClick={[Function]}
                  rel="alternate"
                >
                  latviešu
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/lt/how-it-works"
                  hrefLang="lt"
                  lang="lt"
                  onClick={[Function]}
                  rel="alternate"
                >
                  lietuvių
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/hu/how-it-works"
                  hrefLang="hu"
                  lang="hu"
                  onClick={[Function]}
                  rel="alternate"
                >
                  magyar
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/mt/how-it-works"
                  hrefLang="mt"
                  lang="mt"
                  onClick={[Function]}
                  rel="alternate"
                >
                  Malti
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/nl/how-it-works"
                  hrefLang="nl"
                  lang="nl"
                  onClick={[Function]}
                  rel="alternate"
                >
                  Nederlands
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/pl/how-it-works"
                  hrefLang="pl"
                  lang="pl"
                  onClick={[Function]}
                  rel="alternate"
                >
                  polski
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/pt/how-it-works"
                  hrefLang="pt"
                  lang="pt"
                  onClick={[Function]}
                  rel="alternate"
                >
                  português
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/ro/how-it-works"
                  hrefLang="ro"
                  lang="ro"
                  onClick={[Function]}
                  rel="alternate"
                >
                  română
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/sk/how-it-works"
                  hrefLang="sk"
                  lang="sk"
                  onClick={[Function]}
                  rel="alternate"
                >
                  slovenčina
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/sl/how-it-works"
                  hrefLang="sl"
                  lang="sl"
                  onClick={[Function]}
                  rel="alternate"
                >
                  slovenščina
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/fi/how-it-works"
                  hrefLang="fi"
                  lang="fi"
                  onClick={[Function]}
                  rel="alternate"
                >
                  suomi
                </a>
              </li>
              <li
                className="ecl-language-list__item "
              >
                <a
                  className="ecl-language-list__link ecl-link ecl-link--standalone ecl-link"
                  href="/sv/how-it-works"
                  hrefLang="sv"
                  lang="sv"
                  onClick={[Function]}
                  rel="alternate"
                >
                  svenska
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    data-focus-guard={true}
    style={
      Object {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={0}
  />,
]
`;
